classcastexception updating buildconfig

occurs using the latest version of the plugin, and no kubernetes namespace set (emptystring).

```
INFO: Updating BuildConfig in namespace: pipelineproject with name: sample-pipeline
May 24, 2016 10:00:09 PM io.fabric8.jenkins.openshiftsync.PipelineJobListener upsertBuildConfigForJob
WARNING: Failed to update BuildConfig: NamespaceName{pipelineproject:sample-pipeline}. io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:57)
	at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.replace(HasMetadataOperation.java:112)
	at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.replace(HasMetadataOperation.java:32)
	at io.fabric8.jenkins.openshiftsync.PipelineJobListener.upsertBuildConfigForJob(PipelineJobListener.java:117)
	at io.fabric8.jenkins.openshiftsync.PipelineJobListener.upsertItem(PipelineJobListener.java:93)
	at io.fabric8.jenkins.openshiftsync.PipelineJobListener.onCreated(PipelineJobListener.java:62)
	at hudson.model.listeners.ItemListener$2.apply(ItemListener.java:186)
	at hudson.model.listeners.ItemListener$2.apply(ItemListener.java:184)
	at hudson.model.listeners.ItemListener.forAll(ItemListener.java:167)
	at hudson.model.listeners.ItemListener.fireOnCreated(ItemListener.java:184)
	at hudson.model.ItemGroupMixIn.createProjectFromXML(ItemGroupMixIn.java:280)
	at jenkins.model.Jenkins.createProjectFromXML(Jenkins.java:3105)
	at io.fabric8.jenkins.openshiftsync.BuildConfigWatcher$1.call(BuildConfigWatcher.java:115)
	at io.fabric8.jenkins.openshiftsync.BuildConfigWatcher$1.call(BuildConfigWatcher.java:99)
	at hudson.security.ACL.impersonate(ACL.java:213)
	at io.fabric8.jenkins.openshiftsync.BuildConfigWatcher.upsertJob(BuildConfigWatcher.java:99)
	at io.fabric8.jenkins.openshiftsync.BuildConfigWatcher.onInitialBuildConfigs(BuildConfigWatcher.java:68)
	at io.fabric8.jenkins.openshiftsync.GlobalPluginConfiguration$1.run(GlobalPluginConfiguration.java:172)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.replace(HasMetadataOperation.java:93)
	... 23 more
Caused by: java.lang.ClassCastException: io.fabric8.openshift.api.model.BuildTriggerPolicy cannot be cast to io.fabric8.openshift.api.model.BuildTriggerPolicy
	at io.fabric8.openshift.api.model.BuildConfigSpecFluentImpl.withTriggers(BuildConfigSpecFluentImpl.java:156)
	at io.fabric8.openshift.api.model.BuildConfigSpecBuilder.<init>(BuildConfigSpecBuilder.java:42)
	at io.fabric8.openshift.api.model.BuildConfigFluentImpl.withSpec(BuildConfigFluentImpl.java:63)
	at io.fabric8.openshift.api.model.BuildConfigBuilder.<init>(BuildConfigBuilder.java:32)
	at io.fabric8.openshift.api.model.DoneableBuildConfig.<init>(DoneableBuildConfig.java:20)
	... 28 more

May 24, 2016 10:00:09 PM io.fabric8.jenkins.openshiftsync.BuildConfigWatcher$1 call
```

--- a/src/main/java/io/fabric8/jenkins/openshiftsync/BuildCause.java
+++ b/src/main/java/io/fabric8/jenkins/openshiftsync/BuildCause.java
@@ -16,41 +16,84 @@
 package io.fabric8.jenkins.openshiftsync;
 
 import hudson.model.Cause;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
 import io.fabric8.openshift.api.model.Build;
-import io.fabric8.openshift.api.model.BuildSource;
-import io.fabric8.openshift.api.model.BuildSpec;
+import io.fabric8.openshift.api.model.GitBuildSource;
 import org.apache.commons.lang.StringUtils;
 
 public class BuildCause extends Cause {
 
-  private final Build build;
+  private String uid;
 
-  public BuildCause(Build build) {
-    this.build = build;
+  private String namespace;
+
+  private String name;
+
+  private String gitUri;
+
+  private String commit;
+
+  public BuildCause(String uid, String namespace, String name, String gitUri, String commit) {
+    this.uid = uid;
+    this.namespace = namespace;
+    this.name = name;
+    this.gitUri = gitUri;
+    this.commit = commit;
   }
 
-  public Build getBuild() {
-    return build;
+  public BuildCause(Build build) {
+    if (build == null || build.getMetadata() == null) {
+      return;
+    }
+    ObjectMeta meta = build.getMetadata();
+    uid = meta.getUid();
+    namespace = meta.getNamespace();
+    name = meta.getName();
+
+    if (build.getSpec() != null) {
+      if (build.getSpec().getSource() != null && build.getSpec().getSource().getGit() != null) {
+        GitBuildSource git = build.getSpec().getSource().getGit();
+        gitUri = git.getUri();
+      }
+
+      if (build.getSpec().getRevision() != null && build.getSpec().getRevision().getGit() != null) {
+        commit = build.getSpec().getRevision().getGit().getCommit();
+      }
+    }
   }
 
   @Override
   public String getShortDescription() {
     StringBuilder sb = new StringBuilder("OpenShift Build ")
-      .append(build.getMetadata().getNamespace()).append("/").append(build.getMetadata().getName());
+      .append(namespace).append("/").append(name);
 
-    BuildSpec spec = build.getSpec();
-    if (spec != null) {
-      BuildSource source = spec.getSource();
-      if (source != null && source.getGit() != null && StringUtils.isNotBlank(source.getGit().getUri())) {
-        sb.append(" from ").append(source.getGit().getUri());
-        if (spec.getRevision() != null
-          && spec.getRevision().getGit() != null
-          && StringUtils.isNotBlank(spec.getRevision().getGit().getCommit())) {
-          sb.append(", commit ").append(spec.getRevision().getGit().getCommit());
-        }
+    if (StringUtils.isNotBlank(gitUri)) {
+      sb.append(" from ").append(gitUri);
+      if (StringUtils.isNotBlank(commit)) {
+        sb.append(", commit ").append(commit);
       }
     }
 
     return sb.toString();
+  }
+
+  public String getUid() {
+    return uid;
+  }
+
+  public String getNamespace() {
+    return namespace;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public String getGitUri() {
+    return gitUri;
+  }
+
+  public String getCommit() {
+    return commit;
   }
 }

--- a/src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigProjectProperty.java
+++ b/src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigProjectProperty.java
@@ -21,10 +21,9 @@ import hudson.model.JobProperty;
 import hudson.model.JobPropertyDescriptor;
 import io.fabric8.openshift.api.model.BuildConfig;
 import jenkins.model.ParameterizedJobMixIn;
-import org.apache.commons.lang.StringUtils;
 import org.kohsuke.stapler.DataBoundConstructor;
 
-import static io.fabric8.jenkins.openshiftsync.JenkinsUtils.xstream2;
+import static io.fabric8.jenkins.openshiftsync.OpenShiftUtils.getOpenShiftClient;
 
 /**
  * Stores the OpenShift Build Config related project properties.
@@ -35,29 +34,38 @@ import static io.fabric8.jenkins.openshiftsync.JenkinsUtils.xstream2;
  */
 public class BuildConfigProjectProperty extends JobProperty<Job<?, ?>> {
 
-  // The build config this job relates to.
-  private BuildConfig buildConfig;
+  // The build config uid this job relates to.
+  private String uid;
+
+  private String namespace;
+
+  private String name;
 
   @DataBoundConstructor
-  public BuildConfigProjectProperty(String buildConfigAsXML) {
-    if (StringUtils.isNotBlank(buildConfigAsXML)) {
-      this.buildConfig = (BuildConfig) xstream2().fromXML(buildConfigAsXML);
-    }
+  public BuildConfigProjectProperty(String namespace, String name, String uid) {
+    this.namespace = namespace;
+    this.name = name;
+    this.uid = uid;
   }
 
-  public BuildConfigProjectProperty(BuildConfig buildConfig) {
-    this.buildConfig = buildConfig;
+  public String getBuildConfigUid() {
+    return uid;
   }
 
   public BuildConfig getBuildConfig() {
-    return buildConfig;
+    BuildConfig bc = getOpenShiftClient().buildConfigs().inNamespace(namespace).withName(name).get();
+    if (bc != null && bc.getMetadata().getUid().equals(uid)) {
+      return bc;
+    }
+    return null;
   }
 
-  public String getBuildConfigAsXML() {
-    if (buildConfig == null) {
-      return null;
-    }
-    return xstream2().toXML(buildConfig);
+  public String getName() {
+    return name;
+  }
+
+  public String getNamespace() {
+    return namespace;
   }
 
   @Extension

--- a/src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigWatcher.java
+++ b/src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigWatcher.java
@@ -18,7 +18,7 @@ package io.fabric8.jenkins.openshiftsync;
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 import hudson.model.Job;
 import hudson.security.ACL;
-import io.fabric8.kubernetes.api.model.Status;
+import hudson.util.XStream2;
 import io.fabric8.kubernetes.client.KubernetesClientException;
 import io.fabric8.kubernetes.client.Watch;
 import io.fabric8.kubernetes.client.Watcher;
@@ -40,7 +40,6 @@ import java.util.logging.Level;
 import java.util.logging.Logger;
 
 import static io.fabric8.jenkins.openshiftsync.BuildConfigToJobMapper.mapBuildConfigToJob;
-import static io.fabric8.jenkins.openshiftsync.JenkinsUtils.xstream2;
 import static io.fabric8.jenkins.openshiftsync.OpenShiftUtils.getOpenShiftClient;
 import static io.fabric8.jenkins.openshiftsync.OpenShiftUtils.isJenkinsBuildConfig;
 import static io.fabric8.jenkins.openshiftsync.OpenShiftUtils.parseResourceVersion;
@@ -169,9 +168,15 @@ public class BuildConfigWatcher implements Watcher<BuildConfig> {
             return null;
           }
 
-          jobFromBuildConfig.addProperty(new BuildConfigProjectProperty(buildConfig));
+          jobFromBuildConfig.addProperty(
+            new BuildConfigProjectProperty(
+              buildConfig.getMetadata().getNamespace(),
+              buildConfig.getMetadata().getName(),
+              buildConfig.getMetadata().getUid()
+            )
+          );
 
-          InputStream jobStream = new StringInputStream(xstream2().toXML(jobFromBuildConfig));
+          InputStream jobStream = new StringInputStream(new XStream2().toXML(jobFromBuildConfig));
 
           Jenkins jenkins = Jenkins.getInstance();
           Job job = BuildTrigger.getDscp().getJobFromBuildConfigUid(buildConfig.getMetadata().getUid());

--- a/src/main/java/io/fabric8/jenkins/openshiftsync/BuildSyncRunListener.java
+++ b/src/main/java/io/fabric8/jenkins/openshiftsync/BuildSyncRunListener.java
@@ -26,9 +26,6 @@ import hudson.model.listeners.RunListener;
 import io.fabric8.openshift.api.model.Build;
 import jenkins.util.Timer;
 import org.jenkinsci.plugins.workflow.job.WorkflowRun;
-import org.joda.time.DateTime;
-import org.joda.time.format.DateTimeFormatter;
-import org.joda.time.format.ISODateTimeFormat;
 import org.kohsuke.stapler.DataBoundConstructor;
 
 import javax.annotation.Nonnull;
@@ -112,11 +109,9 @@ public class BuildSyncRunListener extends RunListener<Run> {
     if (shouldPollRun(run)) {
       try {
         BuildCause cause = (BuildCause) run.getCause(BuildCause.class);
-        if (cause != null && cause.getBuild() != null) {
+        if (cause != null) {
           // TODO This should be a link to the OpenShift console.
-          run.setDescription(
-            cause.getBuild().getMetadata().getNamespace() + "/" + cause.getBuild().getMetadata().getName()
-          );
+          run.setDescription(cause.getShortDescription());
         }
       } catch (IOException e) {
         logger.log(Level.WARNING, "Cannot set build description: " + e);
@@ -195,16 +190,12 @@ public class BuildSyncRunListener extends RunListener<Run> {
     }
 
     BuildCause cause = (BuildCause) run.getCause(BuildCause.class);
-    if (cause == null || cause.getBuild() == null) {
+    if (cause == null) {
       return;
     }
 
-    final Build build = cause.getBuild();
-
     String rootUrl = OpenShiftUtils.getJenkinsURL(getOpenShiftClient(), defaultNamespace);
     String logsUrl = joinPaths(rootUrl, run.getUrl(), "/consoleText");
-
-    String name = build.getMetadata().getName();
 
     String phase = runToBuildPhase(run);
 
@@ -220,12 +211,8 @@ public class BuildSyncRunListener extends RunListener<Run> {
       }
     }
 
-    if (logger.isLoggable(Level.FINE)) {
-      logger.fine("generated build in namespace " + defaultNamespace + " with name: " + name + " phase: " + build.getStatus().getPhase() + " data: " + build);
-    }
-
-    logger.info("replacing build in namespace " + defaultNamespace + " with name: " + name + " phase: " + build.getStatus().getPhase());
-    getOpenShiftClient().builds().inNamespace(defaultNamespace).withName(name).edit()
+    logger.info("Patching build in namespace " + cause.getNamespace() + " with name: " + cause.getName() + " phase: " + phase);
+    getOpenShiftClient().builds().inNamespace(cause.getNamespace()).withName(cause.getName()).edit()
       .editMetadata()
         .addToAnnotations(ANNOTATION_JENKINS_STATUS_JSON, json)
         .addToAnnotations(ANNOTATION_JENKINS_BUILD_URI, run.getUrl())

--- a/src/main/java/io/fabric8/jenkins/openshiftsync/BuildTrigger.java
+++ b/src/main/java/io/fabric8/jenkins/openshiftsync/BuildTrigger.java
@@ -48,7 +48,7 @@ public class BuildTrigger extends Trigger<Job<?, ?>> {
       return;
     }
 
-    DESCRIPTOR.addBuildConfigTrigger(buildConfigProjectProperty.getBuildConfig().getMetadata().getUid(), super.job);
+    DESCRIPTOR.addBuildConfigTrigger(buildConfigProjectProperty.getBuildConfigUid(), super.job);
   }
 
   @Override

--- a/src/main/java/io/fabric8/jenkins/openshiftsync/JenkinsUtils.java
+++ b/src/main/java/io/fabric8/jenkins/openshiftsync/JenkinsUtils.java
@@ -21,7 +21,6 @@ import hudson.model.Job;
 import hudson.model.Queue;
 import hudson.model.Run;
 import hudson.model.TopLevelItem;
-import hudson.util.XStream2;
 import io.fabric8.openshift.api.model.Build;
 import io.fabric8.openshift.api.model.BuildConfig;
 import jenkins.model.Jenkins;
@@ -91,7 +90,7 @@ public class JenkinsUtils {
       Queue buildQueue = jenkins.getQueue();
       for (Queue.Item item : buildQueue.getItems()) {
         for (Cause cause : item.getCauses()) {
-          if (cause instanceof BuildCause && ((BuildCause) cause).getBuild().getMetadata().getUid().equals(buildUid)) {
+          if (cause instanceof BuildCause && ((BuildCause) cause).getUid().equals(buildUid)) {
             buildQueue.cancel(item);
             cancelOpenShiftBuild(build);
             return;
@@ -102,7 +101,7 @@ public class JenkinsUtils {
         if (obj instanceof WorkflowRun) {
           WorkflowRun b = (WorkflowRun) obj;
           BuildCause cause = b.getCause(BuildCause.class);
-          if (cause != null && cause.getBuild().getMetadata().getUid().equals(buildUid)) {
+          if (cause != null && cause.getUid().equals(buildUid)) {
             Executor e = b.getExecutor();
             if (e != null) {
               e.interrupt(ABORTED);
@@ -113,14 +112,6 @@ public class JenkinsUtils {
       }
       cancelOpenShiftBuild(build);
     }
-  }
-
-  public static XStream2 xstream2() {
-    XStream2 xs = new XStream2();
-    xs.aliasType("OpenShiftBuildConfig", BuildConfig.class);
-    xs.addCompatibilityAlias(BuildConfig.class.getName(), BuildConfig.class);
-    xs.addCompatibilityAlias("OpenShiftBuildConfig", BuildConfig.class);
-    return xs;
   }
 
   public static Job getJobFromBuild(Build build) {

